### ISSUE TITLE
[freebsd] TCP transport not working in 2015.8.0rc3

### ISSUE THREAD

**Issue by cedwards** [saltstack/salt#26364]
In testing the `transport: tcp` option on my FreeBSD systems I'm seeing the following output from my master. This output is repeated until the master is stopped..

```
==> /var/log/salt/master <==
2015-08-16 11:42:10,159 [salt.utils.process                       ][DEBUG   ][787] Started 'salt.transport.tcp.<type 'instancemethod'>._publish_daemon' with pid 4894
2015-08-16 11:42:10,166 [salt.utils.process                       ][INFO    ][787] Process <bound method TCPPubServerChannel._publish_daemon of <salt.transport.tcp.TCPPubServerChannel object at 0x80d307150>> (4894) died with exit status None, restarting...
2015-08-16 11:42:11,175 [salt.utils.process                       ][DEBUG   ][787] Started 'salt.transport.tcp.<type 'instancemethod'>._publish_daemon' with pid 4895
2015-08-16 11:42:11,182 [salt.utils.process                       ][INFO    ][787] Process <bound method TCPPubServerChannel._publish_daemon of <salt.transport.tcp.TCPPubServerChannel object at 0x80d307150>> (4895) died with exit status None, restarting...
```

System details:
```
Salt Version:
           Salt: 2015.8.0rc3

Dependency Versions:
         Jinja2: 2.8
       M2Crypto: 0.22
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.6.0
         Python: 2.7.10 (default, Jul  4 2015, 05:19:16)
           RAET: 0.6.3
        Tornado: 4.2
            ZMQ: 4.1.2
          ioflo: 1.2.4
        libnacl: 1.4.2
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
       pycrypto: 2.6.1
        timelib: 0.2.4

System Versions:
           dist:
        machine: amd64
        release: 10.1-RELEASE-p16
```

**Comment by cedwards** [comment]
...additional info...

`salt-call` appears to work (but only if applying this sysctl setting):

```
sysctl security.jail.allow_raw_sockets=1
```

Targeting from the master returns the error:

```
The salt master could not be contacted. Is master running?
```

Running the master in debug also repeats the following error:

```
emma ~ # salt-master -l debug                                                                                                                                                                    [901/1467]
[DEBUG   ] Reading configuration from /usr/local/etc/salt/master
[DEBUG   ] Including configuration from '/usr/local/etc/salt/master.d/_schedule.conf'
[DEBUG   ] Reading configuration from /usr/local/etc/salt/master.d/_schedule.conf
[DEBUG   ] Including configuration from '/usr/local/etc/salt/master.d/transport.conf'
[DEBUG   ] Reading configuration from /usr/local/etc/salt/master.d/transport.conf
[DEBUG   ] Using cached minion ID from /usr/local/etc/salt/minion_id: emma
[DEBUG   ] Configuration file path: /usr/local/etc/salt/master
[INFO    ] Setting up the Salt Master
[DEBUG   ] Loaded master key: /usr/local/etc/salt/pki/master/master.pem
[INFO    ] Preparing the root key for local communication
[DEBUG   ] Removing stale keyfile: /var/cache/salt/master/.root_key
[DEBUG   ] Created pidfile: /var/run/salt-master.pid
[INFO    ] The salt master is starting up
[DEBUG   ] LazyLoaded roots.envs
[DEBUG   ] Could not LazyLoad roots.init
[INFO    ] salt-master is starting as user 'root'
[INFO    ] Current values for max open files soft/hard setting: 176121/176121
[INFO    ] Creating master process manager
[INFO    ] Creating master maintenance process
[DEBUG   ] Started 'salt.transport.tcp.<type 'instancemethod'>._publish_daemon' with pid 15832
[INFO    ] Creating master event publisher process
Process Process-1:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/local/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python2.7/site-packages/salt/transport/tcp.py", line 611, in _publish_daemon
    pub_server.listen(int(self.opts['publish_port']), address=self.opts['interface'])
  File "/usr/local/lib/python2.7/site-packages/tornado/tcpserver.py", line 127, in listen
    self.add_sockets(sockets)
  File "/usr/local/lib/python2.7/site-packages/tornado/tcpserver.py", line 144, in add_sockets
    io_loop=self.io_loop)
  File "/usr/local/lib/python2.7/site-packages/tornado/netutil.py", line 266, in add_accept_handler
[DEBUG   ] Started 'salt.utils.event.<type 'type'>.EventPublisher' with pid 15833
    io_loop.add_handler(sock, accept_handler, IOLoop.READ)
  File "/usr/local/lib/python2.7/site-packages/tornado/ioloop.py", line 704, in add_handler
    self._impl.register(fd, events | self.ERROR)
  File "/usr/local/lib/python2.7/site-packages/tornado/platform/kqueue.py", line 41, in register
    self._control(fd, events, select.KQ_EV_ADD)
  File "/usr/local/lib/python2.7/site-packages/tornado/platform/kqueue.py", line 63, in _control
    self._kqueue.control([kevent], 0)
OSError: [Errno 9] Bad file descriptor
[DEBUG   ] Started 'salt.master.<type 'type'>.Maintenance' with pid 15834
[INFO    ] Creating master publisher process
[INFO    ] Creating master concache process
[DEBUG   ] ConCache initializing...
[DEBUG   ] ConCache cleaning up
[DEBUG   ] ConCache-Timer started
[DEBUG   ] Started 'salt.utils.master.<type 'type'>.ConnectedCache' with pid 15835
[DEBUG   ] Sleeping for two seconds to let concache rest
...
```

**Comment by cedwards** [comment]
In continued testing I have tried the latest develop branch in a virtualenv and see the same results:

```
(salt.git)emma ~ # salt --versions-report
Salt Version:
           Salt: 2015.8.0-262-gc34c6b9

Dependency Versions:
         Jinja2: 2.8
       M2Crypto: 0.22
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.7.0
         Python: 2.7.10 (default, Jul  4 2015, 05:19:16)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.2
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
       pycrypto: 2.6.1
        timelib: 0.2.4

System Versions:
           dist:
        machine: amd64
        release: 10.1-RELEASE-p16
```

**Comment by jfindlay** [comment]
@cedwards, thanks for the report.  We found some other TCP issues during Beryllium testing.  It will therefore be considered an experimental feature until Boron.

**Comment by cedwards** [comment]
FYI - just tested the `transport: tcp` option in the RC4 release. I now get less and different output, but it's still not functional.

```
[ERROR   ] Future exception was never retrieved: Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/tornado/gen.py", line 230, in wrapper
    yielded = next(result)
  File "/usr/local/lib/python2.7/site-packages/salt/minion.py", line 494, in eval_master
    pub_channel = salt.transport.client.AsyncPubChannel.factory(self.opts, **factory_kwargs)
  File "/usr/local/lib/python2.7/site-packages/salt/transport/client.py", line 159, in factory
    return salt.transport.tcp.AsyncTCPPubChannel(opts, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/salt/transport/tcp.py", line 195, in __init__
    self.io_loop = kwargs['io_loop'] or tornado.ioloop.IOLoop.current()
KeyError: 'io_loop'
```

I realize we're considering it experimental. Just continuing to track the issue here.

**Comment by JaseFace** [comment]
@cedwards thanks for testing these experimental releases.  Before I give it a go, have you tried anything recently?  We are running 2015.8.1 and have started hitting some bottlenecks in ZMQ, looking for some greener pastures.

**Comment by cedwards** [comment]
@JaseFace I just tried again on 2015.8.1 and get the following:

```
saturn salt # salt '*' test.ping
[ERROR   ] Exception in callback <function <lambda> at 0x80c5fc7d0> for <tornado.concurrent.Future object at 0x80c5ed450>
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/tornado/concurrent.py", line 300, in _set_done
    cb(self)
  File "/usr/local/lib/python2.7/site-packages/tornado/ioloop.py", line 584, in <lambda>
    lambda future: self.add_callback(callback, future))
  File "/usr/local/lib/python2.7/site-packages/tornado/ioloop.py", line 914, in add_callback
    raise RuntimeError("IOLoop is closing")
RuntimeError: IOLoop is closing
[ERROR   ] Future exception was never retrieved: StreamClosedError
```

I'd really like to get this working, but I'm not sure where to start in patching for FreeBSD.

**Comment by basepi** [comment]
@cedwards This is still on our radar, we just need to test and try to reproduce. Might be worth trying 2015.8.3 (when it gets released, hopefully next week or the week after) and the new tornado 4.3. I'm not super optimistic either of these things will fix the issues you're seeing, but it's something to try.

**Comment by DmitryKuzmenko** [comment]
Reproduced, investigating.

**Comment by cachedout** [comment]
Great work Dmitry. Looking forward to hearing what you discover.

**Comment by DmitryKuzmenko** [comment]
Tornado io_loop doesn't work properly with bsd fork. Not sure about details, but a simple test with once called IOLoop.current() doesn't work in a new process using it. So any process should not get ioloop before fork or it should create a new one after to work on BSD.

**Comment by cedwards** [comment]
This appears to be resolved in 2016.3.0rc2. I've just confirmed connectivity to a dozen jailed-minions using `transport: tcp`.
